dsl.go

```diff
@@ -528,14 +528,19 @@ type InOrderContext struct {
 	lastInvokedMethodParams []Param
 }
 
-// Matcher ... it is guaranteed that FailureMessage will always be called after Matches
-// so an implementation can save state
-type Matcher interface {
+// ArgumentMatcher can be used to match arguments but not invocation counts.
+type ArgumentMatcher interface {
 	Matches(param Param) bool
-	FailureMessage() string
 	fmt.Stringer
 }
 
+// Matcher can be used to match arguments as well as invocation counts. It is guaranteed that
+// FailureMessage will always be called after Matches so an implementation can save state.
+type Matcher interface {
+	ArgumentMatcher
+	FailureMessage() string
+}
```

Comment:
Let's go all the way here. This should probably be:
```go
type Matcher interface {
	Matches(param Param) bool
	FailureMessage() string
}
```

We never need the Stringer for invocation count matching.

Reply:
I left the `Matcher` interface equivalent to what it was before to avoid breaking backwards compatibility for anybody else using it. I could introduce a `InvocationCountMatcher` interface that looks the way you suggest and make the `Matcher` interface the union of the two. I also think that might make a bit more sense naming wise as the change would make the interface more specialized.

Reply:
Or if backwards compatibility is not really a concern then skip the union interface, i.e. your suggestion plus name change.

Reply:
Ok, so I went ahead and added the `InvocationCountMatcher` ifc and kept the `Matcher` around. If you don't like it I can remove the `Matcher` ifc.